### PR TITLE
Terraform V2

### LINKED ISSUE #882: azdo terraofrm task: update deprecated arm warnings
Hi,

AZDO terraform task (task: TerraformTaskV1@0) for azure has deprecated warnings for some time now.
["arm_subscription_id": [DEPRECATED] `arm_subscription_id` has been replaced by `subscription_id`
["arm_tenant_id": [DEPRECATED] `arm_tenant_id` has been replaced by `tenant_id`
["arm_client_id": [DEPRECATED] `arm_client_id` has been replaced by `client_id`
["arm_client_secret": [DEPRECATED] `arm_client_secret` has been replaced by `client_secret`

https://github.com/microsoft/azure-pipelines-extensions/blob/master/Extensions/Terraform/Src/Tasks/TerraformTaskV1/src/azure-terraform-command-handler.ts
these 4 lines need updating:
this.backendConfig.set('**arm_subscription_id**', tasks.getEndpointDataParameter(backendServiceName, "subscriptionid", true));
this.backendConfig.set('**arm_tenant_id**', tasks.getEndpointAuthorizationParameter(backendServiceName, "tenantid", true));
this.backendConfig.set('**arm_client_id**', tasks.getEndpointAuthorizationParameter(backendServiceName, "serviceprincipalid", true));
this.backendConfig.set('**arm_client_secret**', tasks.getEndpointAuthorizationParameter(backendServiceName, "serviceprincipalkey", true));

what is the process to submitt a pr to get the task updated?

regards 

Richard Marchesi

### PR DESCRIPTION
**Pull request by aristosvo**
> Credits to @hfurubotten for the description

The arm_ prefix on backend configurations has been deprecated back in 2018 and released with terraform v0.12. Updated to match latest documentation from terraform which lists current configuration for terraform v0.12 and v0.13 (v0.14 is around the corner

This is a continuation of PR #770 and #863, but with the updates in a new taks version. Hopefully those will go through with success now.

Related issues:
fixes #882

### CODE REVIEW

**Comment by jikuja** [comment]
I feel like this PR should contain more than one commit:

- First copy all files into the V2 directory as one commit
- Then all the changes divided into multiple commits if needed. Depends on the complexity of the changes.

Otherwise, extra manual work is needed to find the actual changes for debugging and/or auditing purposes.

**Comment by aristosvo** [comment]
Good suggestion! I‘ll rework the PR within a few days, thanks!

**Review by amarkulis — APPROVED** [review]

**Comment by kaovd** [comment]
Bumping this, was literally about to PR it then i realised should check the existing ones first lol.

File: Extensions/Terraform/Src/Tasks/TerraformTaskV2/README.md
Comment (hfurubotten) [inline Extensions/Terraform/Src/Tasks/TerraformTaskV2/README.md:27]: Could perhaps update this text to include that only `terraform>=v0.12` will be supported by the backend configuration.

Comment (aristosvo) [reply Extensions/Terraform/Src/Tasks/TerraformTaskV2/README.md]: Thanks, resolved!

**Comment by kaovd** [comment]
Any updates on this? Seems all ready to be merged.

**Comment by StefanSchoof** [comment]
Terraform just removed the deprecated fields in https://github.com/hashicorp/terraform/pull/28181.
This means with actions of the maintainers this extension will not work with the next major terraform version.

**Comment by aristosvo** [comment]
@AmrutaKawade @sebader Please take this PR into consideration, as mentioned by @StefanSchoof the [next version of Terraform which won't support these fields anymore](https://github.com/hashicorp/terraform/releases/tag/v0.15.0-beta2) is around the corner

**Comment by pavanvamsi3** [comment]
CC: @chshrikh, please check this PR once.

**Comment by DenWin** [comment]
> The arm_ prefix on backend configurations has been deprecated back in 2018 and released with terraform v0.12.

Just saying, but these fields will not be deprecated anymore with Terraform 0.15 - they will be GONE!

**Comment by pavanvamsi3** [comment]
Thanks for that scene @DenWin 🙈 

I'm pinging the owner of this right away, also is this PR tested?

**Comment by pavanvamsi3** [comment]
@chshrikh is looking into it 🌱

**Comment by runemy** [comment]
Please also look this PR: https://github.com/microsoft/azure-pipelines-extensions/pull/852

**Comment by Stephen-T-Kaye** [comment]
I've worked around this for the time being by using the AzurePowerShell@5 task to retrieve the credentials from the service connection and then run `terraform init` from powershell. The following inline script works because of the way the AzurePowerShell tasks invokes the script which means a variable is in scope (`endpointObject`) that contains the required information. The PowerShell script itself can probably be drastically improved but I'm sharing as is in case anyone else would like to be able to use the latest version of Terraform. Personally I'm using it to take advantage of the Azure AD auth so that I can exercise greater control over the storage account used to store the remote state.

```
- task: AzurePowerShell@5
  displayName: Terraform init
  inputs:
    azureSubscription: "[my service connection name]"
    scriptType: inlineScript
    inline: |
      $sa = "[my storage account name]"
      $rg = "[my resource group name]"
      $cn = "terraform-state"
      $ck = "terraform.tfstate"
      $subscriptionId = $endpointObject | Select -Expand Data | Select -ExpandProperty subscriptionId
      $tenantId = $endpointObject | Select -Expand Auth | select -Expand parameters | Select -ExpandProperty tenantid
      $clientId = $endpointObject | Select -Expand Auth | select -Expand parameters | Select -ExpandProperty serviceprincipalid
      $clientSecret = $endpointObject | Select -Expand Auth | select -Expand parameters | Select -ExpandProperty serviceprincipalkey
      $clientSecret = $clientSecret -replace "\$", "``$"
      
      $cmd = "terraform.exe init -reconfigure -backend-config=""use_azuread_auth=true"" -backend-config=""storage_account_name=$sa"" -backend-config=""container_name=$cn"" -backend-config=""key=$ck"" -backend-config=""resource_group_name=$rg"" -backend-config=""subscription_id=$subscriptionId"" -backend-config=""tenant_id=$tenantId"" -backend-config=""client_id=$clientId"" -backend-config=""client_secret=$clientSecret"""
      Write-Host $cmd
      $cmd | Invoke-Expression
      if ($LASTEXITCODE -gt 0) {
        exit $LASTEXITCODE
      }
    failOnStderr: true
    azurePowerShellVersion: latestVersion
```

**Comment by StefanSchoof** [comment]
@Stephen-T-Kaye Did you try to `addSpnToEnvironment` of the Azure cli task (https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/deploy/azure-cli?view=azure-devops)? This may a more stable solution.

**Comment by alxy** [comment]
Yes, here is a [blog entry](https://cloudarchitected.com/2021/02/using-terraform-in-azure-pipelines-without-extensions/) describing the setup without extension. (TLS certificate seems to be invalid currently of that site, but here is the relevant part):

```yaml
- task: AzureCLI@1
  displayName: Terraform credentials
  inputs:
    azureSubscription: My-Service-Connection
    scriptLocation: inlineScript
    inlineScript: |
      set -eu  # fail on error
      subscriptionId=$(az account show --query id -o tsv)
      echo "##vso[task.setvariable variable=ARM_CLIENT_ID]$servicePrincipalId"
      echo "##vso[task.setvariable variable=ARM_CLIENT_SECRET;issecret=true]$servicePrincipalKey"
      echo "##vso[task.setvariable variable=ARM_SUBSCRIPTION_ID]$subscriptionId"
      echo "##vso[task.setvariable variable=ARM_TENANT_ID]$tenantId"
    addSpnToEnvironment: true

- task: AzureCLI@1
  displayName: Terraform init
  inputs:
    azureSubscription: My-Service-Connection
    scriptLocation: inlineScript
    inlineScript: |
      set -eux  # fail on error
      subscriptionId=$(az account show --query id -o tsv)
      terraform init \
        -backend-config=storage_account_name=${{ parameters.TerraformBackendStorageAccount }} \
        -backend-config=container_name=${{ parameters.TerraformBackendStorageContainer }} \
        -backend-config=key=${{ parameters.environment }}.tfstate \
        -backend-config=resource_group_name=${{ parameters.TerraformBackendResourceGroup }} \
        -backend-config=subscription_id="$(ARM_SUBSCRIPTION_ID)" \
        -backend-config=tenant_id="$(ARM_TENANT_ID)" \
        -backend-config=client_id="$(ARM_CLIENT_ID)" \
        -backend-config=client_secret="$(ARM_CLIENT_SECRET)"    workingDirectory: ${{ parameters.TerraformDirectory }}
    addSpnToEnvironment: true
```

Keep in mind that there might be two different service connections required, one to access the backend state file, and another one to actually deploy the resources to a potentially different Azure subscription. In this example, the same service principal (belonging to `My-Service-Connection`) is used for both, which is a common setup.

/cc @algattik

**Comment by Stephen-T-Kaye** [comment]
@alxy @StefanSchoof Thanks for the pointers. Unfortunately I'm using self-hosted agents that don't currently have the Azure CLI installed. I will probably change that as I believe the `addSpnToEnvironment` would be more stable.

Thanks for providing the excerpt as my corp proxy prevents my access due to the invalid cert. I noticed in the second task the `addSpnToEnvironment: true` is not required nor is the line `subscriptionId=$(az account show --query id -o tsv)` as the script in the second task only depends on the environment variables exposed via `task.setvariable` in the first task. 

However, If one only wishes to work around the issue with `init` but still use the extension for the other terraform commands, it should be possible without exposing the credentials to any later tasks with the following variation of the above excerpt.
 
```yaml
- task: AzureCLI@1
  displayName: Terraform init
  inputs:
    azureSubscription: My-Service-Connection
    scriptLocation: inlineScript
    inlineScript: |
      set -eux  # fail on error
      subscriptionId=$(az account show --query id -o tsv)
      terraform init \
        -backend-config=storage_account_name="mystorageacc" \
        -backend-config=container_name="terraformcontainer" \
        -backend-config=key="terraform_state.tfstate" \
        -backend-config=resource_group_name="my-backend-rg" \
        -backend-config=subscription_id="$subscriptionId" \
        -backend-config=tenant_id="$tenantId" \
        -backend-config=client_id="$servicePrincipalId" \
        -backend-config=client_secret="$servicePrincipalKey" 
    addSpnToEnvironment: true
```

**Comment by aristosvo** [comment]
[0.15.0 is released..](https://github.com/hashicorp/terraform/releases/tag/v0.15.0)

Please @pavanvamsi3 @chshrikh take action

**Comment by kaovd** [comment]
Is the assignee able to be moved to someone else or is this completely blocked on one person atm?

**Comment by CaptainStealthy** [comment]
@pavanvamsi3 @harshiniks @shashban @chshrikh @Roshan-Kumar-Microsoft @RoshanKumarMicrosoft @RoopeshNair 

This issue directly blocks us from moving to Terraform 0.15. Please review this PR as soon as possible.

I have to say, Microsoft's lack of care/attention for this repository is getting pretty ridiculous...

**Comment by DenWin** [comment]
To add insult to injury, these fields are deprecatex since 2018!

**Comment by amarkulis** [comment]
@CaptainStealthy @DenWin What happened is Microsoft acquired github in 2018 and they shifted all Azure Devops engineers and resources to work on Github because that is their strategic future. So Azure Devops improvements and maintenance have fallen by the wayside. You can see that there is a sharp drop off of all Azure Devops repo management around 2018.

**Comment by pavanvamsi3** [comment]
Hey folks, I'm following up with the owner once again.

There have been some even more critical issues due to which they couldn't take a look at this I suppose.

**Comment by kaovd** [comment]
@pavanvamsi3 any updates? another weeks gone by...

**Comment by terryjrjoyce** [comment]
When will this be merged so we can start using v15?

**Comment by timja** [comment]
> When will this be merged so we can start using v15?

you can use this instead if you wish:
https://github.com/charleszipp/azure-pipelines-tasks-terraform

**Comment by artur-carvalho** [comment]
This Init fail is really annoying, is there any plan to merge this?

**Comment by sajkat** [comment]
@pavanvamsi3 this is blocking the adoption of 0.15. Can we get some traction on this?

**Comment by StefanSchoof** [comment]
And since the images now have terraform 0.15 installed (see https://github.com/actions/virtual-environments/releases/tag/ubuntu18%2F20210419.1) Pipelines without an install step start to fail.

**Comment by ennui93** [comment]
Terraform 0.15.1 showed up in our Azure DevOps pipelines today as it's now the default offering from https://apt.releases.hashicorp.com, so we encountered the backend configuration issue for the first time.

Can we prioritize this pull request? This is going to start hitting a lot of folks now.

**Comment by kaovd** [comment]
@hfurubotten Is there no one else this PR can be reassigned to from Microsoft than would be relevant to give an @? This is very quick to approve and seems to be going no where at the moment

**Comment by DenWin** [comment]
While I agree, this was neglected far to long, this is by far not an easy fix.
For all of you now wanting to use 0.15 there are no arm_-something variables anymore - sure!
But there is a second side of this coin, as all that are still using 0.11, there only are the arm_-something variables and none of the new ones.

That said - it is now 14 days, or 2.5 years - depending on how you wanna see it - please fix it!

**Comment by StefanSchoof** [comment]
That is reason why this is version 2 of the extension. You can run version 1 with old terraform and Version 2 with new versions

**Comment by DenWin** [comment]
> That is reason why this is version 2 of the extension. You can run version 1 with old terraform and Version 2 with new versions

Well, I was referencing "This is very quick to approve" With V2 there are changes in 153 files, that is not easy either!

**Comment by kaovd** [comment]
> While I agree, this was neglected far to long, this is by far not an easy fix.
> For all of you now wanting to use 0.15 there are no arm_-something variables anymore - sure!
> But there is a second side of this coin, as all that are still using 0.11, there only are the arm_-something variables and none of the new ones.
> 
> That said - it is now 14 days, or 2.5 years - depending on how you wanna see it - please fix it!

Its in a separated version provider so this does not break v1 deployments

**Comment by kaovd** [comment]
> > That is reason why this is version 2 of the extension. You can run version 1 with old terraform and Version 2 with new versions
> 
> Well, I was referencing "This is very quick to approve" With V2 there are changes in 153 files, that is not easy either!

Bit misleading to say 153, v1 was duplicated and there are 8 changes in practice

**Comment by Stephen-T-Kaye** [comment]
I recently encountered an issue that would also be present in this version. The [azuread provider](https://registry.terraform.io/providers/hashicorp/azuread/latest/docs) still uses the `ARM_` prefix for environment variables when [configuring the service principal](https://registry.terraform.io/providers/hashicorp/azuread/latest/docs/guides/service_principal_client_secret#configuring-the-service-principal-in-terraform).
We could add a small change to set the new environment variables and the `ARM_` prefixed ones by including [both of these blocks in the V2 commit](https://github.com/microsoft/azure-pipelines-extensions/pull/889/commits/03a55ce549a8060fe362329f15ad08a477f64fb0#diff-6d045688d326a1a62da46a01c1c9b8e61776305ce6f7280eb87fbcd4d9b8d793L17)

**Comment by DenWin** [comment]
> I recently encountered an issue that would also be present in this version. The [azuread provider](https://registry.terraform.io/providers/hashicorp/azuread/latest/docs) still uses the `ARM_` prefix for environment variables when [configuring the service principal](https://registry.terraform.io/providers/hashicorp/azuread/latest/docs/guides/service_principal_client_secret#configuring-the-service-principal-in-terraform).
> We could add a small change to set the new environment variables and the `ARM_` prefixed ones by including [both of these blocks in the V2 commit](https://github.com/microsoft/azure-pipelines-extensions/pull/889/commits/03a55ce549a8060fe362329f15ad08a477f64fb0#diff-6d045688d326a1a62da46a01c1c9b8e61776305ce6f7280eb87fbcd4d9b8d793L17)

Your sure, you aren't confusing Environment variables with backend variables?
One is indeed (even with Terraform 0.15):
```bash
$ export ARM_CLIENT_ID="00000000-0000-0000-0000-000000000000"
$ export ARM_CLIENT_SECRET="00000000-0000-0000-0000-000000000000"
$ export ARM_TENANT_ID="10000000-2000-3000-4000-500000000000"
```
and the other is backend-variables:
```bash
terraform init -backend-config=arm_subscription_id=00000000-0000-0000-0000-000000000000 -backend-config=arm_tenant_id=*** -backend-config=arm_client_id=*** -backend-config=arm_client_secret=***
```

**Comment by Stephen-T-Kaye** [comment]
@DenWin is correct, it was mishap in my workaround that highlighted how the extension is a little presumptuous and sets environment variables used by specific providers (albeit sometimes helpful). I'd confused the two for a moment.

**Review by rari2012 — APPROVED** [review]
well done

**Comment by kaovd** [comment]
Im gonna echo the final verdict in #852, this extension seems flat out dead. Time to make use of the workarounds or forked repos.

**Comment by algattik** [comment]
See [my blog for some insights](https://cloudarchitected.com/2021/02/using-terraform-in-azure-pipelines-without-extensions/), my team mostly either uses scripting or [Charles Zipp's extension](https://marketplace.visualstudio.com/items?itemName=charleszipp.azure-pipelines-tasks-terraform).

**Comment by Theuno** [comment]
> See [my blog for some insights](https://cloudarchitected.com/2021/02/using-terraform-in-azure-pipelines-without-extensions/), my team mostly either uses scripting or [Charles Zipp's extension](https://marketplace.visualstudio.com/items?itemName=charleszipp.azure-pipelines-tasks-terraform).

Of course that is a solution, but not all organisations allow or prefer that solution. A Microsoft governed extension is then preferred.
Too bad Microsoft doesn't support the community on this one. Effectively putting their platforms a step back in the end...

**Comment by aristosvo** [comment]
> See [my blog for some insights](https://cloudarchitected.com/2021/02/using-terraform-in-azure-pipelines-without-extensions/), my team mostly either uses scripting or [Charles Zipp's extension](https://marketplace.visualstudio.com/items?itemName=charleszipp.azure-pipelines-tasks-terraform).

The repo of Charles is offline.., who knows. Maybe Microsoft is into it after all 😍

**Comment by timja** [comment]
err @charleszipp what's going on with that ^^?

**Comment by marchesir** [comment]
hi,
terraform 0.15 will now break the extension:
https://www.terraform.io/upgrade-guides/0-15.html
althougth terraform init -reconfigure will tell terraform to replace old names with new ones.

so any updated when v2 will be released of the extension?

regards

**Comment by ajklotz** [comment]
What's going on here? We can't use Microsoft's terraform release task with 0.15.3 because I'm getting the errors about invalid backend configuration arguments.

**Comment by kaovd** [comment]
> What's going on here? We can't use Microsoft's terraform release task with 0.15.3 because I'm getting the errors about invalid backend configuration arguments. 

MS dont care enough to review this, so its either DIY (see above), use the lovely @charleszipp extension, or cry.

**Comment by CaptainStealthy** [comment]
@pavanvamsi3 @chshrikh Any update?

Given that @algattik works for another team at Microsoft and is using alternative workarounds says a lot.

I'm going to give this until I get some free time to upgrade from TF 0.14 to 0.15, and if this issue hasn't moved by then, I will be looking at workarounds as well.

**Comment by kaovd** [comment]
@CaptainStealthy you can use @Stephen-T-Kaye  's solution and then after init you can use the plan apply etc commands in the terraform v1 extension like normal.

**Comment by marchesir** [comment]
hi,
looks like msft doesnt care about the extension as all theie engery is on github actions. so my team looks like we will just use pure tf as the extension is a real pain to just update a few env variables.

regards

**Comment by ajklotz** [comment]
> > > See [my blog for some insights](https://cloudarchitected.com/2021/02/using-terraform-in-azure-pipelines-without-extensions/), my team mostly either uses scripting or [Charles Zipp's extension](https://marketplace.visualstudio.com/items?itemName=charleszipp.azure-pipelines-tasks-terraform).
> > 
> > 
> > The repo of Charles is offline.., who knows. Maybe Microsoft is into it after all 😍
> 
> FYI, my repo will be coming back online this week. I had to take it private temporarily to resolve an internal issue. The current version already accommodates 0.15 so that could be another potential workaround until this is resolved.

Do you have a PR or commit where you fixed it to make it work? I was trying to confirm

**Comment by charleszipp** [comment]
@ajklotz i do not because that commit was made before my repo was recreated so the history is no longer available. However, the code for how I handle this can be viewed from [this file](https://github.com/charleszipp/azure-pipelines-tasks-terraform/blob/4983947161d5d4b28334529843c4760164cf3bdc/tasks/terraform-cli/src/backends/azurerm.ts#L23). Basically at version 0.12 and after the configuration variables removed the `arm_` prefix. So my extension checks the version of terraform being used and sets the backend configuration accordingly.

The e2e tests also verify this is working by using 0.15 which can be seen in this pipeline yaml [here](https://github.com/charleszipp/azure-pipelines-tasks-terraform/blob/main/pipelines/test/smoke_test.yml)

**Comment by ajklotz** [comment]
> @ajklotz i do not because that commit was made before my repo was recreated so the history is no longer available. However, the code for how I handle this can be viewed from [this file](https://github.com/charleszipp/azure-pipelines-tasks-terraform/blob/4983947161d5d4b28334529843c4760164cf3bdc/tasks/terraform-cli/src/backends/azurerm.ts#L23). Basically at version 0.12 and after the configuration variables removed the `arm_` prefix. So my extension checks the version of terraform being used and sets the backend configuration accordingly.
> 
> The e2e tests also verify this is working by using 0.15 which can be seen in this pipeline yaml [here](https://github.com/charleszipp/azure-pipelines-tasks-terraform/blob/main/pipelines/test/smoke_test.yml)

Oh, i just noticed this is a build pipeline thing only. Thats unfortunate. I was wanting to use this in the release pipeline but it doesn't look like that is supported.

**Comment by charleszipp** [comment]
@ajklotz the extension works within classic ui editor for releases as well. See screen shot below
image

**Comment by ajklotz** [comment]
> @ajklotz the extension works within classic ui editor for releases as well. See screen shot below. In fact, this is how the extension itself was tested and published to the marketplace until earlier this year.
> image

My mistake! I didn't notice that the iconography was different. The marketplace icon is white, then the tasks you get are purple like the other MS tasks :D

**Comment by jonsamwell** [comment]
Any update?

**Comment by aristosvo** [comment]
Fixed by #958

**Pull request closed without merging**
